src/anomalib/models/components/base/post_processing.py

```diff
@@ -0,0 +1,118 @@
+from abc import ABC, abstractmethod
+
+import torch
+from lightning import LightningModule, Trainer
+from lightning.pytorch import Callback
+from torch import nn
+
+from anomalib.dataclasses import Batch, InferenceBatch
+from anomalib.metrics import F1AdaptiveThreshold, MinMax
+
+
+class PostProcessor(nn.Module, Callback, ABC):
+    @abstractmethod
+    def forward(self, pred_score, anomaly_map):
+        """Funcional forward method for post-processing."""
+
+    @abstractmethod
+    def post_process_batch(self, batch: Batch):
+        """Post-process the predictions."""
+
+
+class OneClassPostProcessor(PostProcessor):
+    """Default post-processor for one-class anomaly detection."""
```

Comment:
I would argue that normalization and thresholding are both post-procesing operations, since they are applied to the raw model predictions in order to enhance their interpretability.

In the previous design we had separate callbacks for normalization, thresholding and post-processing, which was confusing (normalization and thresholding can also be seen as post-processing steps) and prone to bugs (the different callbacks were dependent on each other's outputs). 

The new design and terminology simplifies this. All post-processing operations are now collected in a single class, so we don't need to pass some variables around. For example, the image and pixel threshold are now directly accessible during both thresholding and normalization. 

Some of the simple operations that were previously part of the post processing callback are now handled by the `__post_init__` of the dataclass. I feel it is warranted to do this outside of the post processor, because it's nothing more than just computing some additional fields based on the available data (e.g. image score from anomaly map by taking the max pixel value).

> I am in favour of merging Normalization, post-processing, thresholding, and metrics computation in a single class as they all depend on each other and in specific order.

I personally feel that metric computation should still have its own separate callback. The metric computation is independent of any post-processing operations, as the computation steps are the same with or without post-processing. Including it in the post-processor would lead to code duplication and violate the single responsibility principle.

> We will need to figure out how to make this configurable, and handle different thresholding/normalization approaches.

The most straightforward way to configure this is to have the user define the post-processor class path and init args as part of the engine arguments. The engine can then set the post-processor attribute of the model or use the model's default one if nothing is specified by the user. This would be achievable with minor changes to `AnomalyModule` and `Engine`. 

In general, I think we should keep our default post-processor class simple. In practice we will use min-max normalization and adaptive thresholding in 99% of the use-cases, so maybe we should not add unnecessary complexity to this class by adding different thresholding/normalization approaches. When a user really needs some alternative post-processing steps, they can easily implement a custom post-processor and enable it through the trainer.